robot-server/robot_server/protocols/protocol_store.py

```diff
@@ -199,6 +199,14 @@ def get_all(self) -> List[ProtocolResource]:
             for r in all_sql_resources
         ]
 
+    def get_id_by_hash(self, hash: str) -> Optional[str]:
+        """Get all protocol hashes keyed by protocol id."""
+        try:
+            p_source = next(p for p in self.get_all() if p.source.content_hash == hash)
+            return p_source.protocol_id
+        except StopIteration:
+            return None
```

Comment:
This is just a matter of style, but I'd personally just write this procedurally, to avoid having to do the quirky `except StopIteration` thing.

```suggestion
        for protocol in self.get_all():
            if p.source.content_hash == hash:
                return p.protocol_id
        return None
```

-----

A separate point, if you want to keep the `next`/`StopIteration`-based implementation:

The naming here might be a little confusing. `p_source` is a `ProtocolResource`, not a `ProtocolSource` as you might expect from its name. (Our names `ProtocolResource` and `ProtocolSource` are themselves not very good. Sorry.)